internal/mesh/internal/controllers/sidecarproxy/controller_test.go

```diff
@@ -256,114 +260,259 @@ func (suite *meshControllerTestSuite) TestReconcile_ExistingProxyStateTemplate_N
-		Write(suite.T(), suite.client.ResourceServiceClient)
+		serviceRef := resource.ReferenceToString(resource.Reference(suite.apiService.Id, ""))
+		suite.client.WaitForStatusCondition(t, webDestinations.Id, ControllerName,
+			status.ConditionMeshProtocolFound(serviceRef))
+
+		// We should also get a new web proxy template resource as this destination should be added again.
+		webProxyStateTemplate = suite.client.WaitForNewVersion(t, webProxyStateTemplateID, webProxyStateTemplate.Version)
 
-	suite.client.WaitForStatusCondition(suite.T(), webDestinations.Id, ControllerName,
-		status.ConditionMeshProtocolFound(serviceRef))
+		requireExplicitDestinationsFound(t, "api", webProxyStateTemplate)
+	})
 
-	// We should also get a new web proxy template resource as this destination should be added again.
-	webProxyStateTemplate = suite.client.WaitForNewVersion(suite.T(), webProxyStateTemplateID, webProxyStateTemplate.Version)
+	testutil.RunStep(suite.T(), "delete the proxy state template and check re-generation", func(t *testing.T) {
+		// Delete the proxy state template resource and check that it gets regenerated.
+		suite.runtime.Logger.Trace("deleting web proxy")
+		_, err := suite.client.Delete(suite.ctx, &pbresource.DeleteRequest{Id: webProxyStateTemplateID})
+		require.NoError(suite.T(), err)
 
-	// Delete the proxy state template resource and check that it gets regenerated.
-	_, err := suite.client.Delete(suite.ctx, &pbresource.DeleteRequest{Id: webProxyStateTemplateID})
-	require.NoError(suite.T(), err)
+		webProxyStateTemplate = suite.client.WaitForNewVersion(suite.T(), webProxyStateTemplateID, webProxyStateTemplate.Version)
+		requireExplicitDestinationsFound(t, "api", webProxyStateTemplate)
+	})
+
+	testutil.RunStep(suite.T(), "add implicit upstream and enable tproxy", func(t *testing.T) {
+		// Delete explicit destinations resource.
+		suite.runtime.Logger.Trace("deleting web destinations")
+		_, err := suite.client.Delete(suite.ctx, &pbresource.DeleteRequest{Id: webDestinations.Id})
+		require.NoError(t, err)
+
+		webProxyStateTemplate = suite.client.WaitForNewVersion(suite.T(), webProxyStateTemplateID, webProxyStateTemplate.Version)
 
-	suite.client.WaitForNewVersion(suite.T(), webProxyStateTemplateID, webProxyStateTemplate.Version)
+		// Enable transparent proxy for the web proxy.
+		resourcetest.Resource(types.ProxyConfigurationType, "proxy-config").
+			WithData(t, &pbmesh.ProxyConfiguration{
+				Workloads: &pbcatalog.WorkloadSelector{
+					Prefixes: []string{"web"},
+				},
+				DynamicConfig: &pbmesh.DynamicConfig{
+					Mode: pbmesh.ProxyMode_PROXY_MODE_TRANSPARENT,
+					TransparentProxy: &pbmesh.TransparentProxy{
+						OutboundListenerPort: 15001,
+					},
+				},
+			}).Write(t, suite.client)
+
+		webProxyStateTemplate = suite.client.WaitForNewVersion(suite.T(), webProxyStateTemplateID, webProxyStateTemplate.Version)
+
+		requireImplicitDestinationsFound(t, "api", webProxyStateTemplate)
```

Comment:
Why does this show up if the `Service` object for `api` never had an exposed `mesh` port in the fixtures? I would have expected that the only services eligible for implicit upstreams would be those that exposed a mesh port on the service itself?

Reply:
See comment on `data_fetcher.go`

Reply:
The ports on the api get updated to have mesh in the previous step